### ISSUE TITLE
Overenthusiastic updating the timestamp on a contact

### ISSUE THREAD

**Issue by conoroneill** [jobisoft/DAV-4-TbSync#49]
## Your environment

TbSync version: 0.7.22
Thunderbird version: 60.2.1, Linux (Xubuntu)

[Y] Yes, I have installed the latest available (beta) version from 
https://github.com/jobisoft/TbSync/releases 
and my issue is not yet fixed, I can still reproduce it.

The plug-in appears to be updating the contacts every time they are _used_, not just when the contact's details are changed.

## Expected behavior
A contact should only be marked with a new timestamp, and synchronised via CardDav, if the user has changed any of the contact's fields.

## Actual behavior
Every time that a contact is _used_, eg if you compose an email to that contact, then the timestamp is updated, and therefore the contact is re-synchronised via CardDav.

## Steps to reproduce
Link an account via CardDav, and investigate the synchronisation behaviour.

To help resolving your issue, enable debug logging (TbSync Account Manager -> Help) and send me the debug.log via e-mail (use the title of your issue as subject of the email).

**Comment by jobisoft** [comment]
Thats a bummer. The TB addressbook itself is updating the timestamp on each usage. I just get notified that "something" changed (not possibloe to know what). I could try to compare the card I am about to send with the version last received from server and if they only differ in timestamp, do not send it.

Is this a real issue for you in terms of traffic?

**Comment by conoroneill** [comment]
It's not a major issue for me, it just feels a bit wrong.
I use Kolab as my CardDav provider, and they store the address book entries as pseudo-emails in an IMAP server. These emails are being updated (via TbSync) each time I use an address book entry within Thunderbird, which is what I noticed. It would be cleaner if the records were not updated 'unnecessarily'.

**Comment by jobisoft** [comment]
I "fixed" it by suppressing the PUT request, if the send vCard would be identical to the one currently stored on the server. TbSync will still inform the user about unsynced changes.  I changed the notification from "warning" to "info" for that in general.

Please update to beta by installing all XPI modules in the beta folder (mixing of stable and beta is not possible):
https://tbsync.jobisoft.de/beta

**Comment by jobisoft** [comment]
Can I close this ad fixed?

**Comment by conoroneill** [comment]
Er, not yet, I haven't had a chance to try this yet. Thanks for your work on this. I'll try to have a look this week.

**Comment by conoroneill** [comment]
Yes thanks, this appears to have fixed the issue.

**Comment by conoroneill** [comment]
Unfortunately this behaviour seems to have reappeared, I think this restarted as of TbSync v1.7.

**Comment by jobisoft** [comment]
I think this is a single event for each contact after the update. 
1. With your current setup, it should happen with each contact only once.
2. If you disable your account and re-enable it running 1.7, it should not happen again.

Can you confirm that? If 1) is not true, is 2) true?

**Comment by conoroneill** [comment]
Thanks. I had already upgraded to 1.7 and tried disabling/re-enabling.
It is definitely happening multiple times; I have tried:

- sending an email to a test account
- synchronising
- verified that the account had been updated at the far end
- sending another email to a test account
- synchronising
- verified that the account had been updated again at the far end

Therefore 2) is not true.

**Comment by jobisoft** [comment]
I added some more debug logging for this case in the latest beta. Please switch to beta or update your addons, if you are already on beta.

If this happens again, the log will contain the following line:

```
Card has been modified!
```

followed by the old and the new card. You get access to the log by creating a bug report. 

You can either send the entire log to me, or open the attachment of the generated email and look for yourself. If you find the difference between the two cards (which triggered the PUT command) you can just send me that.

**Comment by conoroneill** [comment]
I finally had some time to look into this again.
I use 'KolabNow' as my DAV server (https://kolabnow.com/). This is what I'm seeing as the delta:

** Mon Mar 11 2019 20:15:34 GMT+0000 (GMT) **
[currentCard] : BEGIN:VCARD
VERSION:3.0
PRODID:-//Kolab//iRony DAV Server 0.4-dev//Sabre//Sabre VObject 3.5.3//EN
UID:a6582a20-91f1-11e6-8286-5254ea544a7a
FN:Test Puddle
N:Puddle;Test;;;
EMAIL;TYPE=INTERNET,OTHER:test@xxxxxx.co.uk
REV:20190311T201231Z
END:VCARD

** Mon Mar 11 2019 20:15:34 GMT+0000 (GMT) **
[newCard] : BEGIN:VCARD
VERSION:3.0
PRODID:-//Kolab//iRony DAV Server 0.4-dev//Sabre//Sabre VObject 3.5.3//EN
UID:a6582a20-91f1-11e6-8286-5254ea544a7a
FN:Test Puddle
N:Puddle;Test;;;
EMAIL;TYPE=INTERNET;TYPE=OTHER:test@xxxxxx.co.uk
REV:20190311T201231Z
END:VCARD

One end is saying
EMAIL;TYPE=INTERNET,OTHER:test@xxxxxx.co.uk
but the other end is saying
EMAIL;TYPE=INTERNET;TYPE=OTHER:test@xxxxxx.co.uk

**Comment by jobisoft** [comment]
Is your server not updating the card? I mean, I send him the new card and await his acknowledgement and still the next time the same contact is „used“ in Thunderbird and thus marked as changed your server is still holding the not-updated-card?

In other words, the next time you use this card and TbSync is doing the check if so something changed, you get the exact same debug output? I cannot confirm that with any other dav server :-(

**Comment by conoroneill** [comment]
I have repeated the test, and yes, the Kolab server is sending back exactly the same syntax the second time:
EMAIL;TYPE=INTERNET,OTHER:test@xxxxxx.co.uk
when you've previously sent it:
EMAIL;TYPE=INTERNET;TYPE=OTHER:test@xxxxxx.co.uk

If I _export_ the VCard directly from my Kolab server (manually, not via the API); it sends
EMAIL;TYPE=INTERNET;TYPE=OTHER:test@xxxxxx.co.uk

If I export the address book from _Thunderbird's_ address book, as .vcf, I get:
email;internet:test@xxxxxx.co.uk

And if I export my contact from my Android phone, which is also synced to the same DAV server, I get:
EMAIL;TYPE=X-null:test@xxxxxx.co.uk

So obviously, there's a lot of variation of what is being considered as allowed in a VCard format.

I also note that in January/February this year, Thunderbird's Address book UI changed significantly, most notably in the way that it handles multiple email addresses or multiple phone numbers; I wonder if this is also a factor.

I'll try properly looking into the VCard definition(s), and also contacting Kolab and see whether they think that they may have an issue themselves.

**Comment by jobisoft** [comment]
> I also note that in January/February this year, Thunderbird's Address book UI changed significantly, most notably in the way that it handles multiple email addresses or multiple phone numbers; I wonder if this is also a factor.

I did that :-) And that is exactly where this issue comes from.

> I'll try properly looking into the VCard definition(s), and also contacting Kolab and see whether they think that they may have an issue themselves.

I will check first, to make sure I am not doing something wrong. Currently I assume, your server is not doing a 1-to-1 compare of the VCARDS but of the parsed content and sees no difference and thus ignores the PUT. Which results in thios loop.

**Comment by conoroneill** [comment]
The Kolab server is obviously not completely ignoring the PUT, because it's updating the record in its DB, which is what I can see happening (the vCards are actually stored as IMAP email entries, so I see new entries appearing). But it seems that when it next is asked for the value, it is returning the shorter format.
I've just been looking into the VCard spec, and it seems a bit vague but implies that both formats are valid (ie TYPE=type1,type2, and/or TYPE=type1;TYPE=type2).
https://tools.ietf.org/html/rfc6350#section-3.3

An aside; if you're the person maintaining the Thunderbird address book UI, then thank you. Could you please tell me how to _delete_ an email address (or a phone number) from the contact; at the moment it seems that I just have to blank out the value, but it leaves an empty slot visible (and this empty slot appears in the VCard also).

**Comment by jobisoft** [comment]
> An aside; if you're the person maintaining the Thunderbird address book UI, then thank you.

I added the changed U to TbSync, not to Thunderbird itself. And it is only used for CardDAV address books. If you edit a contact in the Standard address book, you will see the Standard UI

> Could you please tell me how to delete an email address (or a phone number) from the contact; at the moment it seems that I just have to blank out the value, but it leaves an empty slot visible (and this empty slot appears in the VCard also).

That sounds like a bug. I did not add a delete option and intended to use the "blanking" technique, which should remove the entry from the list, after you have saved and closed the edit dialog. Next time you open the edit dialog, the blank entry is still there?

> I've just been looking into the VCard spec, and it seems a bit vague but implies that both formats are valid (ie TYPE=type1,type2, and/or TYPE=type1;TYPE=type2).

It looks like I have to compare the parsed values as well and not just the plain text content of the vCards itself. According to the spec, the server is allowed to do these transformations. So this is indeed a bug in TbSync.

**Comment by cjosw** [comment]
Regarding deleting phone numbers (not email addresses);
Thanks for the info; it appears to be user error on my part. When I double click and delete a phone number, it ends up leaving the initial '+' there. If I also manually delete that, it does delete the phone number as you describe.
Note, though, that a note on the UI says 'The first e-mail address in the list will be used as the primary e-mail address'. It would be helpful if there were up/down buttons to re-order these (obviously they can actually be edited to shuffle them around too). And what actually uses the 'primary' email address; thunderbird's auto-complete seems to show them all anyway?
Finally, what does the star in the right-hand-side of the phone numbers / email addresses do?

**Comment by jobisoft** [comment]
Shuffeling will be added, I try to impl. drag and drop, because I do not want to add too much buttons. WIP.

The star is an additional VCARD type called PREF (preferred), it allows to highlight an entry to be the preferred value (for example if someone has two private emails). Only one email can stand at the top of the list (which implies being prefered) but you can have a PREF for work and a PREF for home and so on.

The primary email is the one used for lists. I just had to follow TB internal structure here. If you look at the tabular view in the address book, you see one email in the first email column and all others in the second. Note: Autocomplete is also still WIP, the list of emails in the second column is not split on autocomplete.

**Comment by jobisoft** [comment]
I updated the vCard generator to be able to enforce the simple type notation. I use that only for the comparison of the (re-generated) old and new card.

I am still sending the vCard with multiple type fields, as before, as I think this is the better choice.

But still, this should fix your problem. Can you update your beta addons and try again?

FYI: This is the changeset:
https://github.com/jobisoft/DAV-4-TbSync/commit/ed640784596a4a5f8750a099bcfce4ebaeab8d56

**Comment by cjosw** [comment]
Yes, from a brief test, this appears to be working. Thank you.

**Comment by jobisoft** [comment]
If you find anything, just reopen the issue.

Do you want to get notified, when i add email/phone number shuffeling to the UI and/ or fix the autocomplete?

**Comment by cjosw** [comment]
Thanks for all the work you're doing on this, and yes please, do notify me when the Contacts UI is improved.

**Comment by jobisoft** [comment]
The current beta now has drag'n'drop support to re-order emails and phones:

![image](https://user-images.githubusercontent.com/5830621/56616632-235f2200-661e-11e9-88f4-e5cb0b0b3025.png)

autocomplete is still pending...
